### PR TITLE
JDK installer on OSX

### PR DESCRIPTION
**Pull request by ndeloof**
This installer allows to download official Oracle JDK and install automatically. It mount DMG bundle and unpack the installer's pkg to only extract the actual JDK.

### CODE REVIEW

**Comment by ndeloof** [comment]
@reviewbybees please have a look, my primary use case is cloudbees DEV@Cloud OSX build slaves.

**Comment by stephenc** [comment]
:+1: from me though I have not tested if it actually works ;-)

File: core/src/main/java/hudson/tools/JDKInstaller.java [hunk 1]
```diff
@@ -281,6 +282,51 @@ I tried to locate exactly how InstallShield parses the arguments (and why it use
...
+
+            exit = launcher.launch()
+                    .cmds("umount", dmg.getRemote())
+                    .stdout(log)
+                    .join();
+            if (exit != 0)
+                throw new AbortException(Messages.JDKInstaller_FailedToInstallJDK(exit));
+
+            // We only want the actual JDK sub-package, which "Payload" is actually a tar.gz archive
+            String payload = pkg.list("jdk*.pkg/Payload")[0].getRemote();
```
Comment (tfennelly) [inline core/src/main/java/hudson/tools/JDKInstaller.java]: Would it make sense to check that `pkg.list("jdk*.pkg/Payload")` has entries before calling `[0].getRemote()`?

File: core/src/main/java/hudson/tools/JDKInstaller.java [hunk 2]
```diff
@@ -281,6 +282,51 @@ I tried to locate exactly how InstallShield parses the arguments (and why it use
...
+                    .stdout(log)
+                    .join();
+            if (exit != 0)
+                throw new AbortException(Messages.JDKInstaller_FailedToInstallJDK(exit));
+
+            parent.child("Contents/Home").moveAllChildrenTo(new FilePath(launcher.getChannel(), expectedLocation));
+            parent.child("Contents").deleteRecursive();
+
+            pkg.deleteRecursive();
+            dmg.deleteRecursive();
```
Comment (tfennelly) [inline core/src/main/java/hudson/tools/JDKInstaller.java]: Can any of the earlier `deleteRecursive()` calls fail with exception, leaving things only half cleaned up? Maybe that doesn't matter but if it does then maybe use nested try/finally blocks to ensure the later `deleteRecursive()` calls happen.

Comment (ndeloof) [inline core/src/main/java/hudson/tools/JDKInstaller.java]: I mimic the windows installer here.
Then intention (afaict) is to leave intermediate files for diagnostic purpose

File: core/src/main/java/hudson/tools/JDKInstaller.java (hunk 1 above)
Comment (ndeloof) [inline core/src/main/java/hudson/tools/JDKInstaller.java]: no file found would be an error anyway, so should trigger an exception :P

File: core/src/main/java/hudson/tools/JDKInstaller.java [hunk 3]
```diff
@@ -281,6 +282,61 @@ I tried to locate exactly how InstallShield parses the arguments (and why it use
...
+                    .cmds("pkgutil", "--expand", installer, pkg.getRemote())
+                    .stdout(log)
+                    .join();
+            if (exit != 0)
+                throw new AbortException(Messages.JDKInstaller_FailedToInstallJDK(exit));
+
+            exit = launcher.launch()
+                    .cmds("umount", dmg.getRemote())
+                    .stdout(log)
+                    .join();
```
Comment (orrc) [inline core/src/main/java/hudson/tools/JDKInstaller.java]: Should the `umount` command be called in every case?

Comment (ndeloof) [inline core/src/main/java/hudson/tools/JDKInstaller.java]: Keeping the DMG mounted let you diagnose installation failure.

**Pull request merged**